File: base-v3/src/main/java/com/adyen/checkout/base/util/AmountFormat.java
/*
 * Copyright (c) 2019 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by caiof on 31/7/2019.
 */

package com.adyen.checkout.base.util;

import androidx.annotation.NonNull;

import com.adyen.checkout.base.model.payments.Amount;
import com.adyen.checkout.core.exception.CheckoutException;
import com.adyen.checkout.core.exception.NoConstructorException;
import com.adyen.checkout.core.log.LogUtil;
import com.adyen.checkout.core.log.Logger;

import java.math.BigDecimal;
import java.util.Currency;
import java.util.Locale;

public final class AmountFormat {
    private static final String TAG = LogUtil.getTag();

    /**
     * Convert an {@link Amount} to its corresponding {@link BigDecimal} value.
     *
     * @param amount The {@link Amount} to be converted.
     * @return A {@link BigDecimal} representation of the {@link Amount}.
     */
    @NonNull
    public static BigDecimal toBigDecimal(@NonNull Amount amount) {
        return toBigDecimal(amount.getValue(), amount.getCurrency());
    }

    /**
     * Convert a value in minor units with the corresponding currency code to its {@link BigDecimal} representation.
     *
     * @param value The value in minor units.
     * @param currencyCode The currency code of the value.
     * @return A {@link BigDecimal} representation.
     */
    @NonNull
    public static BigDecimal toBigDecimal(long value, @NonNull String currencyCode) {
        final int fractionDigits = getFractionDigits(currencyCode);

        return BigDecimal.valueOf(value, fractionDigits);
    }

    private static int getFractionDigits(@NonNull String currencyCode) {
        final String normalizedCurrencyCode = currencyCode.replaceAll("[^A-Z]", "").toUpperCase(Locale.ROOT);

        try {
            final CheckoutCurrency checkoutCurrency = CheckoutCurrency.find(normalizedCurrencyCode);
            return checkoutCurrency.getFractionDigits();
        } catch (CheckoutException e) {
            Logger.e(TAG, normalizedCurrencyCode + " is an unsupported currency. Falling back to information from java.util.Currency.", e);
        }

        try {
            final Currency currency = Currency.getInstance(normalizedCurrencyCode);
            return Math.max(currency.getDefaultFractionDigits(), 0);
        } catch (IllegalArgumentException e) {
            Logger.e(TAG, "Could not determine fraction digits for " + normalizedCurrencyCode, e);
            return 0;
        }
    }

    private AmountFormat() {
        throw new NoConstructorException();
    }
}

File: components-core/src/main/java/com/adyen/checkout/components/util/AmountFormat.kt
/*
 * Copyright (c) 2019 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by caiof on 31/7/2019.
 */
package com.adyen.checkout.components.util

import com.adyen.checkout.components.model.payments.Amount
import com.adyen.checkout.components.util.CheckoutCurrency.Companion.find
import com.adyen.checkout.core.exception.CheckoutException
import com.adyen.checkout.core.log.LogUtil
import com.adyen.checkout.core.log.Logger
import java.math.BigDecimal
import java.util.Currency
import java.util.Locale

object AmountFormat {
    private val TAG = LogUtil.getTag()

    /**
     * Convert an [Amount] to its corresponding [BigDecimal] value.
     *
     * @param amount The [Amount] to be converted.
     * @return A [BigDecimal] representation of the [Amount].
     */
    @JvmStatic
    fun toBigDecimal(amount: Amount): BigDecimal {
        return toBigDecimal(amount.value, amount.currency!!)
    }

    /**
     * Convert a value in minor units with the corresponding currency code to its [BigDecimal] representation.
     *
     * @param value The value in minor units.
     * @param currencyCode The currency code of the value.
     * @return A [BigDecimal] representation.
     */
    private fun toBigDecimal(value: Long, currencyCode: String): BigDecimal {
        val fractionDigits = getFractionDigits(currencyCode)
        return BigDecimal.valueOf(value, fractionDigits)
    }

    private fun getFractionDigits(currencyCode: String): Int {
        val normalizedCurrencyCode = currencyCode.replace("[^A-Z]".toRegex(), "").uppercase(Locale.ROOT)
        try {
            val checkoutCurrency = find(normalizedCurrencyCode)
            return checkoutCurrency.fractionDigits
        } catch (e: CheckoutException) {
            Logger.e(
                tag = TAG,
                msg = "$normalizedCurrencyCode is an unsupported currency. Falling back to information from " +
                    "java.util.Currency.",
                tr = e
            )
        }
        return try {
            val currency = Currency.getInstance(normalizedCurrencyCode)
            currency.defaultFractionDigits.coerceAtLeast(0)
        } catch (e: IllegalArgumentException) {
            Logger.e(TAG, "Could not determine fraction digits for $normalizedCurrencyCode", e)
            0
        }
    }
}
